grakn-graql/src/test/java/ai/grakn/graql/internal/reasoner/QueryCacheTest.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Grakn - A Distributed Semantic Database
+ * Copyright (C) 2016-2018 Grakn Labs Limited
+ *
+ * Grakn is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Grakn is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package ai.grakn.graql.internal.reasoner;
+
+import ai.grakn.GraknTx;
+import ai.grakn.concept.Entity;
+import ai.grakn.graql.admin.Answer;
+import ai.grakn.graql.admin.Conjunction;
+import ai.grakn.graql.admin.Unifier;
+import ai.grakn.graql.admin.VarPatternAdmin;
+import ai.grakn.graql.internal.pattern.Patterns;
+import ai.grakn.graql.internal.query.QueryAnswer;
+import ai.grakn.graql.internal.reasoner.cache.LazyQueryCache;
+import ai.grakn.graql.internal.reasoner.cache.QueryCache;
+import ai.grakn.graql.internal.reasoner.iterator.LazyAnswerIterator;
+import ai.grakn.graql.internal.reasoner.query.QueryAnswers;
+import ai.grakn.graql.internal.reasoner.query.ReasonerAtomicQuery;
+import ai.grakn.graql.internal.reasoner.query.ReasonerQueries;
+import ai.grakn.test.rule.SampleKBContext;
+import ai.grakn.util.GraknTestUtil;
+import com.google.common.collect.ImmutableMap;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import static ai.grakn.graql.Graql.var;
+import static java.util.stream.Collectors.toSet;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assume.assumeTrue;
+
+public class QueryCacheTest {
```

Comment:
I think all of the test names here could do with improvement. 

When I looked at `record_Retrieve` I was able to know what your were testing without looking at the code. 

However, when looking at  `stream_Record_Update_Retrieve` I had to look at the details of the method to know what is going on.